### DISCUSSION TITLE
Inquiry about jmpbuf usage in DynaRun function of box64 /src/dynarec.c

### DISCUSSION THREAD

**Discussion by ShanZeHu** [ptitSeb/box64#974] [Q&A] [answered]
Hello ptitSeb,I've recently been delving into the source code of box64 and came across the use of jmpbuf in the DynaRun function. I observed that jmpbuf seems to be set during each iteration of the while loop, which prompted me to ponder its functionality.
In an endeavor to optimize certain applications that entail substantial computations and frequent memory operations, I contemplated moving the initialization of jmpbuf outside the while loop. My preliminary tests indicated no compilation errors and the program still functioned as intended. Moreover, I noticed potential performance enhancements in specific applications.
I'm keen on understanding whether this modification aligns with the foundational design philosophy of box64. Was there an initial design consideration to reset jmpbuf with every invocation of DynaRun? Additionally, when DynaRun is called upon, is the execution always based on a single dynamically recompiled block? I'm keen to grasp the underlying logic of this setup to determine the soundness of my approach and to ascertain if it's theoretically justified.Thank you for your time and expertise!
![image](https://github.com/ptitSeb/box64/assets/101883141/93f347e8-b62c-4e14-ab23-8cd956617e50)

**Comment by ptitSeb** [comment, marked as answer]
This change would defeat the purpose of those jumpbuf.

First thing to note, is that the jumpbuf is not initialized at every while loop. Only in rare occasion (check the "if" in front of it).
Now, the purpose of this jumpbuff is to provide a "rollback" mecanism when a ssignal is restoring the execution to another address. That way, the execution stack on native side is a sane (because the signal might occur anywhere, and so maybe in some native function call).

**Reply by ShanZeHu** [reply]
Alright, I get it. So if I'm aiming to enhance the performance, which aspects of the dynamic recompilation do you think are worth delving into? Or are there any runtime environments where you see potential for improvement? From my current explorations, the results aren't quite promising.
